### PR TITLE
Themes: Add filter validation middleware

### PR DESCRIPTION
**Pull request by seear**
Addresses #7225

Add a middleware function that checks the filters supplied via the URL to the theme search page at /design.

* Removes invalid filters
* Reorders remaining filters to the [canonical order](https://github.com/Automattic/wp-calypso/blob/e5d252ab300af253c8a54bd05d37f5e3a8799372/client/my-sites/themes/theme-filters.js#L244)
* Redirects if resulting URL is different to original

**To Test**
* Check that filters containing terms not in [this list](https://github.com/Automattic/wp-calypso/blob/e5d252ab300af253c8a54bd05d37f5ehttps://github.com/Automattic/wp-calypso/blob/e5d252ab300af253c8a54bd05d37f5e3a8799372/client/my-sites/themes/theme-filters.js#L244) are removed from the URL
* Chech that filters are reordered to an alpha sort on the [complete _taxonomy:term_ string](https://github.com/Automattic/wp-calypso/blob/e5d252ab300af253c8a54bd05d37f5e3a8799372/client/my-sites/themes/theme-filters.js#L244)
* For example, entering the URL http://calypso.localhost:3000/design/filter/red,nonsense,blue should redirect to http://calypso.localhost:3000/design/filter/blue,red

### CODE REVIEW

**Comment by ockham** [comment]
This doesn't seem to cover the case of only one, invalid, filter, http://calypso.localhost:3000/design/filter/nonsense redirects to http://calypso.localhost:3000/design/filter/ which gives a blank screen (with sidebar) when logged in, and a "No themes found" screen when logged out.

File: client/my-sites/themes/index.web.js [hunk 1]
```diff
@@ -6,7 +6,8 @@ import userFactory from 'lib/user';
 import { makeLayout } from 'controller';
 import { makeNavigation, siteSelection } from 'my-sites/controller';
 import { singleSite, multiSite, loggedOut } from './controller';
-import { getSubjects } from './theme-filters.js';
+import { getSubjects } from './theme-filters';
+import validateFilters from './validate-filters';
```
Comment (seear) [inline client/my-sites/themes/index.web.js]: Trouble is, `./controller.jsx` is isomorphic, whereas this new middleware needs `page.js`.

Really, we want this new middleware to be isomorphic, but we don't have an isomorphic way to handle re-directs yet.

Comment (ockham) [inline client/my-sites/themes/index.web.js]: Yeah :-/ Okay, so let's keep it in there for now, but maybe add a comment at the top of that file to the effect of this discussion :-)

**Comment by ockham** [comment]
Looks good. One minor change request [here](https://github.com/Automattic/wp-calypso/pull/8374#discussion_r81772869), but other than that, ready to :ship:

**Pull request merged**